Use wheels for Dataflow postcommit tests

Abacn opened the pull request:
Fixes #25966
Fixes #26049

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Tests on dataflow actually passed and indeed takes much faster: https://ci-beam.apache.org/job/beam_PostCommit_Python310_PR/38/

However, flink runner test not running due to `[Errno 2] No such file or directory: 'apache-beam-2.47.0.dev0/apache_beam/...` likely a racing condition that the temporary directory gets deleted.

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Run Python 3.7 PostCommit

Abacn commented:
Run Python 3.8 PostCommit

Abacn commented:
Run Python 3.9 PostCommit

Abacn commented:
- Python3.7 PostCommit passed

- Python3.8 PostCommit fails on `:sdks:python:test-suites:dataflow:py38:tensorRTtests`
  due to that the wheel is not installed (get silently ignored) on the custom SDK container at us.gcr.io/apache-beam-testing/python-postcommit-it/tensor_rt:latest. The reason is that The container was custom build (https://github.com/apache/beam/issues/22651#issuecomment-1248024028) and is half year old and used Beam 2.40.0, where the suffix validation for Beam wheel has since changed (#24114).

  Short term solution is that still feed the test with tarball for now. When #22651 is done it should be able to enable wheel sdk_location for tensorRTtests.

- Python3.9 PostCommit fails unrelated:
  ```
  FAILED 
  apache_beam/ml/inference/tensorflow_inference_it_test.py::TensorflowInference::test_tf_imagenet_image_segmentation - 
  AssertionError: 'Background' != 'Military Uniform'
  ```
  this test also fails on latest master: https://ci-beam.apache.org/job/beam_PostCommit_Python39/1630/consoleFull

- Python3.10 PostCommit failed earlier at `apache_beam.ml.inference.sklearn_inference_it_test.SklearnInference.test_sklearn_regression` due to #25968 should fixed on latest master

Abacn commented:
Run Python Dataflow V2 ValidatesRunner

Abacn commented:
The last two commits
- move the flag to be opt-in (and added them to corresponding Jenkins task)
- a temporary commit that enables the flag without the need of run seed job. Will revert before merge.

Python PostCommit has been frequently broken recently due to various code changes, and Jenkins is somewhat unstable today so I am not intended run them again at this moment.

Abacn commented:
Python_Dataframes and PythonLint test failure due to "Cargo, the Rust package manager, is not installed or is not on PATH.". This is most likely a transient issue as it is not seen in every task (testPy38Cython, testPy39Cython failed but testPy37Cython and testPy310Cython passed). Jenkins trigger is not working currently so defer rerun

Abacn commented:
R: @tvalentyn
R: @damccorm

Abacn commented:
Run Python Dataflow V2 ValidatesRunner

Abacn commented:
Python Dataflow Runner V2 VR costs 1h17min: https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2_PR/62/ which is 50 minutes shorter than current cron jobs.

tvalentyn commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -2785,15 +2785,70 @@ class BeamModulePlugin implements Plugin<Project> {
+            }
+            def packageFilename = collection.singleFile.getName()
+            // rename to the suffix accepted by sdks/python/container/boot.go.
+            def renamed = packageFilename.replace(
```
Do we actually have to rename the wheel? I think what matters is that it's compatible with target platform, not the exact name.

tvalentyn replied:
ok I see. Can we build the wheel with a configuration setting way where we wouldn't have to rename it?

Abacn replied:
yeah, current approach is kind of hack. manylinux image is built using cibuildwheel, which involves setting up a docker container to build it. Currently our GitHub Action is doing this: https://github.com/apache/beam/blob/004c7f612a57112ceb56aa04de318244ccd61f0e/.github/workflows/build_wheels.yml#L272 if migrated to GitHub Action this would be. If we have migrated to GitHub Action this will be convenient. 

For now, it will need to create a new gradle task to build manylinux (and persumably others) (instead of in-place change to intallGcpTest). Probably create a gradle task under ":sdks:python" -- this may also help users who build custom SDK so they can easily have wheels instead of a tarball that takes long time to install on Dataflow. Will do.

tvalentyn commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -2785,15 +2785,70 @@ class BeamModulePlugin implements Plugin<Project> {
         dependsOn ':sdks:python:sdist'
+
+        // if host system wheel compatible with dataflow runner
+        def compatibleWithDataflow = ("amd64".equalsIgnoreCase(System.getProperty("os.arch")) &&
```
Another idea is to try to get the wheel that ci builds for us (in fact, we build wheel for every PR), but we don't upload wheels to gs://beam-wheels-staging for in-flight PRs, only for submitted commits. we could potentially change that and upload  or all commits, into a bucket with lifecycle configured. The wheels that we do upload are tagged by commit ID. Sounds like more going that route is more complexity though that just building a wheel when we can. An advantage is: we will use the same wheel we release. Maybe this will be simpler when we move everything to github actions.

Abacn replied:
Yes this is simpler when moving to GitHub Action. The PostCommit can just add a prerequisite step of building wheels.

Currently, build wheel is a GitHub Action workflow (https://github.com/apache/beam/blob/master/.github/workflows/build_wheels.yml) and PostCommits are still Jenkins tasks and cannot make use of the former.

Abacn replied:
entered #26049. Hold on this PR for now until that issue is done

Abacn commented on .test-infra/jenkins/job_PostCommit_Python.groovy:
```diff
@@ -42,6 +42,7 @@ ALL_SUPPORTED_VERSIONS.each { pythonVersion ->
             rootBuildScriptDir(commonJobProperties.checkoutDir)
             tasks(":python${versionSuffix}PostCommit")
             commonJobProperties.setGradleSwitches(delegate)
+            switches("-PsdistReuseWheel")
```
Lesson learned, didn't know about (sdist - source distribution; bdist - build distribution)

Abacn commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -2785,15 +2785,70 @@ class BeamModulePlugin implements Plugin<Project> {
         doLast {
-          def distTarBall = "${pythonRootDir}/build/apache-beam.tar.gz"
+          def packageFile = "${pythonRootDir}/build/apache-beam.tar.gz"
+          if (compatibleWithDataflow) {
```
```suggestion
          if (project.hasProperty('useWheelDistribution')) {
```

reminder for myself, when the PR is ready switch to opt-in

Abacn commented:
Run Python 3.8 PostCommit

Abacn commented:
Run Python 3.10 PostCommit

Abacn commented:
Thanks @tvalentyn for the suggestion. I have rewritten the part of build wheel. Summary of change since last review:

- added `bdistPy{version}{platfoem}` tasks in `:sdks:python` gradle project to build wheels.
- created a 'initializeForDataflowJob' task in test-suites/dataflow/common.gradle handling sdk distribution setup.
- reverted changes to BeamModulePlugin.

PostCommit succeeded. Ready for review again.

Abacn commented on gradle.properties:
```diff
@@ -38,5 +38,8 @@ javaVersion=1.8
+# supported flink versions
 flink_versions=1.12,1.13,1.14,1.15,1.16
+# supported python versions
+python_versions=3.7,3.8,3.9,3.10
```
Added a gradle property here because now `:sdks:python` project need it to define bdist tasks. This also simplifies gradle.properties in python/test-suites a little bit because now it has a single source of truth for supported python version

AnandInguva replied:
Can we use this one instead? https://github.com/apache/beam/blob/dced7738d6ae8c1d558e6505843caae80e2b4fef/.test-infra/jenkins/PythonTestProperties.groovy#L22

Abacn replied:
Thanks for pointing to this. Not aware of it. This code path is for Jenkins only (`.test-infra/jenkins/`); gradle files works independently with Jenkins (gradle tasks can trigger locally). To reduce to duplicate declaration, refer to `gradle.properties` in Jenkins script should work

tvalentyn replied:
there is also [beam](https://github.com/apache/beam/tree/master)/[sdks](https://github.com/apache/beam/tree/master/sdks)/[python](https://github.com/apache/beam/tree/master/sdks/python)/[test-suites](https://github.com/apache/beam/tree/master/sdks/python/test-suites)
/gradle.properties 

Would be good to see if we can these version configurations together

tvalentyn replied:
is https://github.com/apache/beam/blob/master/sdks/python/test-suites/gradle.properties accessible from the context where you need this?

Abacn replied:
The problem is `sdks/python/test-suites/gradle.properties` is not accessible to `sdks/python` gradle project. I understand `test-suites/gradle.properties` as a configuration to manage which python version used in different test. While for bdist it belongs to the python root project (just as sdist). I did not find a source of truth for supported Python versions accessible in the `sdks/python` context.

tvalentyn replied:
ack, thanks for checking. @AnandInguva FYI one more place to add python versions in the future.

Abacn replied:
Thanks, will experiment to read gradle.properties in jenkins/github actions so reduce the duplicate declarations of py version

tvalentyn commented on sdks/python/build.gradle:
```diff
@@ -64,6 +65,49 @@ artifacts {
+          environment CIBW_BUILD: "cp${pyversion}-${idsuffix}"
+          environment CIBW_ENVIRONMENT: "SETUPTOOLS_USE_DISTUTILS=stdlib"
+          environment CIBW_BEFORE_BUILD: "pip install cython numpy && pip install --upgrade setuptools"
+          // note: sync cibuildwheel version with GitHub Action build_wheel
```
lets add a pointer to exact place where this needs to be modified  and maybe make it a inline comment, and do the same in the counterpart where wheels are built.

Abacn replied:
thanks, added to both build.gradle and .yml

Abacn commented:
Doing a rebase, will squash the reviewed commit (that is before 8329ee5)

tvalentyn commented:
Reminder about code freeze: https://lists.apache.org/thread/rpmq9r9tbgrry8yco70mzyt4psyk5m20

tvalentyn commented:
Thank you very much, @Abacn !

Abacn commented:
Run Python 3.7 PostCommit

Abacn commented:
Run Python 3.11 PostCommit

Abacn commented:
Jenkins not triggering. The latest commit just switched to opt-in buildWheels from opt-out (convenient for testing without the need of triggering seed job). Merging based on the signal in 697635a.

at there Python postcommit passed. "pending" tests were actually passed; failed Java tests irrelevant.

Abacn commented:
There are 3 stuck Java test and 1 failed java tests unrelated. Other precommits passed. Merging for now.
